File: sdks/php/src/LicenseManagerClient.php
<?php

declare(strict_types=1);

namespace LicenseManager\Sdk;

/**
 * Cliente para a API de validação do LicenseManager.
 * Encapsula autenticação HMAC-SHA256 e os 4 endpoints de validação.
 *
 * @example
 * $client = new LicenseManagerClient(
 *     'https://licensemanager-api.enzojb.com.br',
 *     'seu-token',
 *     'guid-da-licenca'
 * );
 * $login = $client->login('usuario@empresa.com');
 */
class LicenseManagerClient
{
    private const MAX_RETRIES = 3;

    public function __construct(
        private readonly string $baseUrl,
        private readonly string $token,
        private readonly string $licenseId,
        private readonly int    $timeoutSeconds = 30,
    ) {
        if (empty(trim($baseUrl)))   throw new \InvalidArgumentException('baseUrl é obrigatório');
        if (empty(trim($token)))     throw new \InvalidArgumentException('token é obrigatório');
        if (empty(trim($licenseId))) throw new \InvalidArgumentException('licenseId é obrigatório');

        $this->baseUrl   = rtrim($baseUrl, '/');
        $this->token     = $token;
        // Normaliza GUID para lowercase com hífens — igual ao servidor (idLicenca:D)
        $this->licenseId = preg_match(
            '/^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i',
            $licenseId
        ) ? strtolower($licenseId) : $licenseId;
    }

    // -------------------------------------------------------------------------
    // Endpoints públicos
    // -------------------------------------------------------------------------

    /**
     * Valida login de um usuário numa licença.
     */
    public function login(string $userId): LoginResult
    {
        $data = $this->post('api/validacao/login', [
            'idLicenca'            => $this->licenseId,
            'identificadorUsuario' => $userId,
        ]);

        return new LoginResult(
            authorized: (bool) ($data['autorizado'] ?? false),
            sessionId:  $data['idSessao'] ?? null,
        );
    }

    /**
     * Envia heartbeat para manter a sessão ativa.
     */
    public function heartbeat(string $sessionId): void
    {
        $this->post('api/validacao/heartbeat', [
            'idLicenca' => $this->licenseId,
            'idSessao'  => $sessionId,
        ]);
    }

    /**
     * Encerra a sessão (idempotente).
     */
    public function logout(string $sessionId): void
    {
        $this->post('api/validacao/logout', [
            'idLicenca' => $this->licenseId,
            'idSessao'  => $sessionId,
        ]);
    }

    /**
     * Valida ou registra uma instalação da aplicação cliente.
     */
    public function validateInstallation(string $machineId): InstallationResult
    {
        $data = $this->post('api/validacao/instalacao', [
            'idLicenca'           => $this->licenseId,
            'identificadorMaquina'=> $machineId,
        ]);

        return new InstallationResult(
            authorized:        (bool) ($data['autorizado']    ?? false),
            installationId:    $data['idInstalacao'] ?? null,
            alreadyRegistered: (bool) ($data['jaRegistrada']  ?? false),
        );
    }

    // -------------------------------------------------------------------------
    // Infraestrutura HMAC
    // -------------------------------------------------------------------------

    /**
     * Calcula a assinatura HMAC-SHA256.
     * @internal Público para facilitar testes unitários.
     */
    public function computeSignature(string $licenseId, string $timestamp, string $bodyJson): string
    {
        // Normaliza para lowercase com hífens — igual ao servidor (idLicenca:D)
        if (preg_match('/^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i', $licenseId)) {
            $licenseId = strtolower($licenseId);
        }
        $payload = "{$licenseId}:{$timestamp}:{$bodyJson}";
        return hash_hmac('sha256', $payload, $this->token);
    }

    /**
     * @return array<string, mixed>
     */
    private function post(string $path, array $body): array
    {
        $bodyJson  = json_encode($body, JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES);
        $timestamp = gmdate('Y-m-d\TH:i:s\Z');
        $nonce     = bin2hex(random_bytes(16));
        $signature = $this->computeSignature($this->licenseId, $timestamp, $bodyJson);
        $url       = rtrim($this->baseUrl, '/') . '/' . $path;

        $lastError = null;

        for ($attempt = 1; $attempt <= self::MAX_RETRIES; $attempt++) {
            $ch = curl_init($url);
            curl_setopt_array($ch, [
                CURLOPT_POST           => true,
                CURLOPT_POSTFIELDS     => $bodyJson,
                CURLOPT_RETURNTRANSFER => true,
                CURLOPT_TIMEOUT        => $this->timeoutSeconds,
                CURLOPT_HTTPHEADER     => [
                    'Content-Type: application/json',
                    "X-Token: {$this->token}",
                    "X-Timestamp: {$timestamp}",
                    "X-Nonce: {$nonce}",
                    "X-Signature: {$signature}",
                ],
            ]);

            $responseBody = curl_exec($ch);
            $statusCode   = (int) curl_getinfo($ch, CURLINFO_HTTP_CODE);
            $curlError    = curl_error($ch);
            curl_close($ch);

            if ($curlError !== '') {
                $lastError = $curlError;
                if ($attempt < self::MAX_RETRIES) {
                    sleep((int) pow(2, $attempt));
                    continue;
                }
                throw new LicenseManagerException(0, "Erro de rede: {$curlError}");
            }

            if ($statusCode === 429 || $statusCode >= 500) {
                if ($attempt < self::MAX_RETRIES) {
                    sleep((int) pow(2, $attempt));
                    continue;
                }
            }

            if ($statusCode < 200 || $statusCode >= 300) {
                throw new LicenseManagerException($statusCode, (string) $responseBody);
            }

            if ($statusCode === 204 || $responseBody === '' || $responseBody === false) {
                return [];
            }

            return json_decode((string) $responseBody, true) ?? [];
        }

        throw new LicenseManagerException(0, "Erro de rede: {$lastError}");
    }
}

File: sdks/csharp/LicenseManagerSdk/LicenseManagerClient.cs
using System.Globalization;
using System.Net.Http.Json;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace LicenseManagerSdk;

/// <summary>
/// Cliente para a API de validação do LicenseManager.
/// Encapsula geração de HMAC-SHA256 e os 4 endpoints de validação.
/// </summary>
public sealed class LicenseManagerClient : IDisposable
{
    private readonly HttpClient _http;
    private readonly string _token;
    private readonly string _licenseId;
    private readonly bool _ownsHttpClient;

    private static readonly JsonSerializerOptions JsonOpts = new()
    {
        PropertyNamingPolicy        = JsonNamingPolicy.CamelCase,
        DefaultIgnoreCondition      = JsonIgnoreCondition.WhenWritingNull,
        PropertyNameCaseInsensitive = true,
    };

    /// <summary>
    /// Cria um novo cliente com URL base, token e ID da licença.
    /// </summary>
    /// <param name="baseUrl">URL base da API (ex: https://licensemanager-api.enzojb.com.br)</param>
    /// <param name="token">Token de autenticação da licença (obtido na emissão)</param>
    /// <param name="licenseId">GUID da licença</param>
    /// <param name="httpClient">HttpClient opcional (útil para testes e injeção de dependência)</param>
    public LicenseManagerClient(string baseUrl, string token, string licenseId, HttpClient? httpClient = null)
    {
        if (string.IsNullOrWhiteSpace(baseUrl))   throw new ArgumentException("baseUrl é obrigatório.",   nameof(baseUrl));
        if (string.IsNullOrWhiteSpace(token))     throw new ArgumentException("token é obrigatório.",     nameof(token));
        if (string.IsNullOrWhiteSpace(licenseId)) throw new ArgumentException("licenseId é obrigatório.", nameof(licenseId));

        _token     = token;
        _licenseId = Guid.TryParse(licenseId, out var parsed)
            ? parsed.ToString("D")   // normaliza para lowercase com hífens uma única vez
            : licenseId;

        if (httpClient is not null)
        {
            _http           = httpClient;
            _ownsHttpClient = false;
        }
        else
        {
            _http = new HttpClient
            {
                BaseAddress = new Uri(baseUrl.TrimEnd('/') + "/"),
                Timeout     = TimeSpan.FromSeconds(30),
            };
            _ownsHttpClient = true;
        }
    }

    // -------------------------------------------------------------------------
    // Endpoints públicos
    // -------------------------------------------------------------------------

    /// <summary>Valida login de um usuário numa licença.</summary>
    /// <param name="userId">Identificador único do usuário na aplicação cliente</param>
    /// <param name="ct">Token de cancelamento</param>
    /// <returns>Resultado com <see cref="LoginResult.SessionId"/> em caso de sucesso</returns>
    public async Task<LoginResult> LoginAsync(string userId, CancellationToken ct = default)
    {
        var body = new { IdLicenca = _licenseId, IdentificadorUsuario = userId };
        var response = await SendAsync("api/validacao/login", body, ct).ConfigureAwait(false);
        return await DeserializeAsync<LoginResult>(response, ct).ConfigureAwait(false);
    }

    /// <summary>Envia heartbeat para manter a sessão ativa.</summary>
    /// <param name="sessionId">ID da sessão retornado pelo <see cref="LoginAsync"/></param>
    /// <param name="ct">Token de cancelamento</param>
    public async Task HeartbeatAsync(string sessionId, CancellationToken ct = default)
    {
        var body = new { IdLicenca = _licenseId, IdSessao = sessionId };
        var response = await SendAsync("api/validacao/heartbeat", body, ct).ConfigureAwait(false);
        response.EnsureSuccessStatusCode();
    }

    /// <summary>Encerra a sessão (idempotente).</summary>
    /// <param name="sessionId">ID da sessão a encerrar</param>
    /// <param name="ct">Token de cancelamento</param>
    public async Task LogoutAsync(string sessionId, CancellationToken ct = default)
    {
        var body = new { IdLicenca = _licenseId, IdSessao = sessionId };
        var response = await SendAsync("api/validacao/logout", body, ct).ConfigureAwait(false);
        response.EnsureSuccessStatusCode();
    }

    /// <summary>Valida ou registra uma instalação da aplicação cliente.</summary>
    /// <param name="machineId">Identificador único da máquina/dispositivo</param>
    /// <param name="ct">Token de cancelamento</param>
    public async Task<InstallationResult> ValidateInstallationAsync(string machineId, CancellationToken ct = default)
    {
        var body = new { IdLicenca = _licenseId, IdentificadorMaquina = machineId };
        var response = await SendAsync("api/validacao/instalacao", body, ct).ConfigureAwait(false);
        return await DeserializeAsync<InstallationResult>(response, ct).ConfigureAwait(false);
    }

    // -------------------------------------------------------------------------
    // Infraestrutura HMAC
    // -------------------------------------------------------------------------

    private async Task<HttpResponseMessage> SendAsync(string path, object body, CancellationToken ct)
    {
        var bodyJson  = JsonSerializer.Serialize(body, JsonOpts);

        var attempts = 0;
        while (true)
        {
            attempts++;
            var timestamp = DateTime.UtcNow.ToString("yyyy-MM-ddTHH:mm:ssZ", CultureInfo.InvariantCulture);
            var nonce     = Guid.NewGuid().ToString("N");
            var signature = ComputeSignature(_licenseId, timestamp, bodyJson);

            var request = new HttpRequestMessage(HttpMethod.Post, path)
            {
                Content = new StringContent(bodyJson, Encoding.UTF8, "application/json"),
            };
            request.Headers.Add("X-Token",     _token);
            request.Headers.Add("X-Timestamp", timestamp);
            request.Headers.Add("X-Nonce",     nonce);
            request.Headers.Add("X-Signature", signature);

            HttpResponseMessage response;
            try
            {
                response = await _http.SendAsync(request, ct).ConfigureAwait(false);
            }
            catch (HttpRequestException) when (attempts < 3)
            {
                await Task.Delay(TimeSpan.FromSeconds(Math.Pow(2, attempts)), ct).ConfigureAwait(false);
                continue;
            }

            if (attempts < 3 && ((int)response.StatusCode == 429 || (int)response.StatusCode >= 500))
            {
                await Task.Delay(TimeSpan.FromSeconds(Math.Pow(2, attempts)), ct).ConfigureAwait(false);
                continue;
            }

            if (!response.IsSuccessStatusCode)
                throw new LicenseManagerException(response.StatusCode, await response.Content.ReadAsStringAsync(ct).ConfigureAwait(false));

            return response;
        }
    }

    private string ComputeSignature(string licenseId, string timestamp, string bodyJson)
    {
        // licenseId já foi normalizado no construtor (lowercase com hífens)
        var payload = $"{licenseId}:{timestamp}:{bodyJson}";
        var key     = Encoding.UTF8.GetBytes(_token);
        var data    = Encoding.UTF8.GetBytes(payload);
        var hash    = HMACSHA256.HashData(key, data);
        return Convert.ToHexString(hash).ToLowerInvariant();
    }

    private static async Task<T> DeserializeAsync<T>(HttpResponseMessage response, CancellationToken ct)
    {
        var result = await response.Content.ReadFromJsonAsync<T>(JsonOpts, ct).ConfigureAwait(false);
        return result ?? throw new LicenseManagerException(response.StatusCode, "Resposta vazia da API.");
    }

    /// <inheritdoc/>
    public void Dispose()
    {
        if (_ownsHttpClient)
            _http.Dispose();
    }
}

// -------------------------------------------------------------------------
// Modelos de resposta
// -------------------------------------------------------------------------

/// <summary>Resultado do endpoint de login.</summary>
public sealed record LoginResult(
    [property: JsonPropertyName("autorizado")] bool Authorized,
    [property: JsonPropertyName("idSessao")]   string? SessionId);

/// <summary>Resultado do endpoint de validação de instalação.</summary>
public sealed record InstallationResult(
    [property: JsonPropertyName("autorizado")]    bool Authorized,
    [property: JsonPropertyName("idInstalacao")]  string? InstallationId,
    [property: JsonPropertyName("jaRegistrada")]  bool AlreadyRegistered);

// -------------------------------------------------------------------------
// Exceção
// -------------------------------------------------------------------------

/// <summary>Exceção lançada quando a API retorna um erro HTTP.</summary>
public sealed class LicenseManagerException : Exception
{
    /// <summary>Código HTTP de status retornado pela API.</summary>
    public System.Net.HttpStatusCode StatusCode { get; }

    /// <summary>Corpo da resposta de erro.</summary>
    public string ResponseBody { get; }

    /// <summary>Cria uma nova instância com o status e corpo da resposta.</summary>
    /// <param name="statusCode">Código HTTP de status</param>
    /// <param name="responseBody">Corpo da resposta</param>
    public LicenseManagerException(System.Net.HttpStatusCode statusCode, string responseBody)
        : base($"LicenseManager API error {(int)statusCode}: {responseBody}")
    {
        StatusCode   = statusCode;
        ResponseBody = responseBody;
    }
}
